typing/typemod.ml

```diff
@@ -1018,6 +1019,25 @@ and apply_modalities_module_type env modalities = function
       Mty_signature sg
   | (Mty_functor _ | Mty_alias _) as mty -> mty
 
+let loc_of_modes (modes : mode loc list) : Location.t option =
+  (* CR zqian: [Parsetree.modes] should be a record with a field that is
+  the location of the whole modes string. *)
+  let rec loc_end_of_modes (head : mode loc) = function
+    | [] -> head.loc.loc_end
+    | head' :: rest -> loc_end_of_modes head' rest
+  in
+  match modes with
+  | [] -> None
+  | head :: rest ->
+    let loc_start = head.loc.loc_start in
+    let loc_end = loc_end_of_modes head rest in
+    Some {loc_start; loc_end; loc_ghost=false}
+
+let check_no_modal_modules ~env modes =
```

Comment:
Probably need to check a language extension here.

Reply:
Currently we just outright don't support this by raising a type error; we will add the check when we support this feature.